File: script/gemini-web-debug/window.mjs
const OMITTED = "System: [Earlier conversation omitted to fit Gemini Web request limit.]";

export function fit(messages, rendered, limit) {
  if (bytes(rendered) <= limit) return rendered;
  const prefix = messages.findIndex(message => !["system", "developer"].includes(message.role));
  const fixed = bytes(rendered.slice(0, Math.max(prefix, 0))) + Buffer.byteLength(OMITTED) + 2;
  let available = Math.max(0, limit - fixed);
  const starts = turnStarts(messages, Math.max(prefix, 0));
  let selected = rendered.length;
  for (let index = starts.length - 2; index >= 0; index -= 1) {
    const size = bytes(rendered.slice(starts[index], starts[index + 1])) + 1;
    if (size > available) break;
    available -= size;
    selected = starts[index];
  }
  const latest = starts.at(-2) ?? rendered.length - 1;
  const tail = selected < rendered.length ? rendered.slice(selected) : [rendered[latest]];
  return [...rendered.slice(0, Math.max(prefix, 0)), OMITTED, ...tail];
}

function turnStarts(messages, prefix) {
  const starts = messages.flatMap((message, index) =>
    index >= prefix && message.role === "user" ? [index] : []);
  if (starts[0] !== prefix) starts.unshift(prefix);
  starts.push(messages.length);
  return starts;
}

function bytes(items) {
  return items.reduce((total, item) => total + Buffer.byteLength(item), 0)
    + Math.max(0, items.length - 1);
}

File: src/provider/gemini_web/prompt/window.rs
//! Turn-safe request bounding for the undocumented Gemini Web endpoint.

use crate::provider::{Message, Role};

mod oversized;
mod turns;

pub(super) const MAX_BYTES: usize = 256 * 1024;
const OMITTED: &str = "System: [Earlier conversation omitted to fit Gemini Web request limit.]";

pub(super) fn fit(messages: &[Message], rendered: Vec<String>, limit: usize) -> Vec<String> {
    if joined_len(&rendered) <= limit {
        return rendered;
    }
    let prefix = messages
        .iter()
        .take_while(|message| matches!(message.role, Role::System | Role::Developer))
        .count();
    let starts = turns::starts(messages);
    let fixed = joined_len(&rendered[..prefix]) + OMITTED.len() + 2;
    if fixed > limit {
        return oversized::fit(&rendered, prefix, &starts, limit, OMITTED);
    }
    let mut available = limit.saturating_sub(fixed);
    let mut selected = rendered.len();
    for range in starts.windows(2).rev() {
        let size = joined_len(&rendered[range[0]..range[1]]) + 1;
        if size > available {
            break;
        }
        available -= size;
        selected = range[0];
    }
    let tail = if selected < rendered.len() {
        rendered[selected..].to_vec()
    } else {
        turns::latest_user(&rendered, &starts, available)
    };
    let mut output = rendered[..prefix].to_vec();
    output.push(OMITTED.to_string());
    output.extend(tail);
    output
}

fn joined_len(items: &[String]) -> usize {
    items.iter().map(String::len).sum::<usize>() + items.len().saturating_sub(1)
}
